index/miner/lotusidx/metrics.go

```diff
@@ -0,0 +1,29 @@
+package lotusidx
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/global"
+	"go.opentelemetry.io/otel/unit"
+)
+
+var (
+	onchainSubindex = attribute.Key("subindex").String("onchain")
+	metaSubindex    = attribute.Key("subindex").String("meta")
+)
+
+func (mi *Index) initMetrics() {
+	meter := global.Meter("powergate")
+
+	_ = metric.Must(meter).NewFloat64ValueObserver("powergate.index.miner.progress", mi.progressValueObserver, metric.WithDescription("Miner index refresh progress"), metric.WithUnit(unit.Dimensionless))
+	mi.meterRefreshDuration = metric.Must(meter).NewInt64ValueRecorder("powergate.index.miner.refresh.duration", metric.WithDescription("Refresh duration"), metric.WithUnit(unit.Milliseconds))
+}
+
+func (mi *Index) progressValueObserver(ctx context.Context, result metric.Float64ObserverResult) {
+	mi.metricLock.Lock()
+	defer mi.metricLock.Unlock()
+	result.Observe(mi.onchainProgress, onchainSubindex)
+	result.Observe(mi.metaProgress, metaSubindex)
```

Comment:
This metric is defined as:
```
_ = metric.Must(meter).NewFloat64ValueObserver("powergate.index.ask.progress", ai.progressValueObserver, metric.WithDescription("Ask index refresh progress"), metric.WithUnit(unit.Dimensionless))
```
So what happens in the `NewXXXXObserver` style of metrics, is that the OpenTelemetry library will call the callback (`progressValueObserver`) on defined intervals.
This has some benefits of not needing to record quickly changing metrics, and also registering a ton of observations within the same timestamp.

In this case, since the update progress is a gauge, I let this metric be updated on every collection interval.

There are quite a few things to learn, and keep improving. I think [this](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#interpretation) is the best reference as to understand how different measurements work. I still need to give a full read, but that's the best source of documentation for the metrics API.